### PR TITLE
dw-dma: change dw_dma_stop for byt and cht

### PR DESCRIPTION
**Pull request by tlauda**
Changes implementation of dw_dma_stop function for
baytrail and cherrytrail platforms. On this platforms
we cannot simply disable channel during transfer,
because it will hang DMA. Previously we were waiting
for transfer finish, which could take up to one period
size (1 ms in most cases). This patch uses suspend and
drain FIFO mechanism, which lowers time needed to stop
DMA channel to steady ~0.34 ms.

Signed-off-by: Tomasz Lauda <tomasz.lauda@linux.intel.com>

### CODE REVIEW

**Comment by tlauda** [comment]
@ranj063 @juimonen With this patch ipc timeout on trigger stop is still there. Could you try increasing the timeout for baytrail and cherrytrail and check if it fixes the issue?

File: src/drivers/dw/dma.c [hunk 1]
```diff
@@ -639,13 +639,19 @@ static int dw_dma_stop(struct dma *dma, int channel)
 
 	spin_lock_irq(&dma->lock, flags);
 
-	ret = poll_for_register_delay(dma_base(dma) + DW_DMA_CHAN_EN,
-				      CHAN_MASK(channel), 0,
```
Comment (lyakh) [inline src/drivers/dw/dma.c]: Do I understand it correctly, that previously this function wasn't disabling the DMA channel actively at all, it was waiting for the enable bit to go to 0, presumably, as triggered by some other process, e.g. by dw_dma_verify_block(), right? But that function only exists if DW_USE_HW_LLI is defined, and it is not defined on BYT and CHT. So, in fact I don't see what would be disabling the channel previously in this function. Does this mean, that this call was always running on a timeout? Were we getting "ewt" errors? The current timeout on BYT is defined to 1333us. Or was that bit cleared in some other way?

File: src/drivers/dw/dma.c [hunk 2]
```diff
@@ -639,13 +639,19 @@ static int dw_dma_stop(struct dma *dma, int channel)
...
+	dw_write(dma, DW_CFG_LOW(channel), chan->cfg_lo | DW_CFG_CH_SUSPEND |
+		 DW_CFG_CH_DRAIN);
```
Comment (lyakh) [inline src/drivers/dw/dma.c]: not sure we need to OR with .cfg_lo above, but it shouldn't hurt either.

**Comment by juimonen** [comment]
@tlauda what's the issue number, I've never tested/experiences this?

File: src/drivers/dw/dma.c (hunk 1 above)
Comment (tlauda) [reply src/drivers/dw/dma.c]: Non cAVS platform don't have support for hardware linked list, so linked list items are manually reloaded in function dw_dma_chan_reload_lli and also this function enables the channel again (which is disabled after every lli transfer). So yes, we were just waiting for the transfer to be over with the next DMA irq, which usually took from 0.5 ms to 0.8 ms, but it really depends on when you stop the stream.

**Comment by tlauda** [comment]
@juimonen You've mentioned it here: https://github.com/thesofproject/sof/issues/1208#issuecomment-479357875.

File: src/drivers/dw/dma.c (hunk 1 above)
Comment (lyakh) [reply src/drivers/dw/dma.c]: Do you know how the channel is disabled after each transfer? Automatically or is there a write to the CHAN_EN register somewhere that I'm not seeing?

Comment (tlauda) [reply src/drivers/dw/dma.c]: Automatically.

**Comment by juimonen** [comment]
@tlauda aah ok, I've seen this before quite often, but testing couple of days with verbose tracing on it has been quite rare... we should have a separate issue about this, because now we have the 1) topology period count change, 2) the copy_lli change in resume and this mixed up.

Anyway I can test this, takes little bit time as I'm now in nocodec kernel...

File: src/drivers/dw/dma.c (hunk 1 above)
Comment (lyakh) [reply src/drivers/dw/dma.c]: thanks! Unrelated to this patch, but it would be nice to add a comment explaining that to the driver.

**Comment by xiulipan** [comment]
@tlauda @ranj063 
Today I check with @xun2z  about the IPC timeout issue on BYT.
Please see update in https://github.com/thesofproject/linux/issues/808

File: src/drivers/dw/dma.c [hunk 3]
```diff
@@ -639,13 +639,19 @@ static int dw_dma_stop(struct dma *dma, int channel)
 
 	spin_lock_irq(&dma->lock, flags);
 
-	ret = poll_for_register_delay(dma_base(dma) + DW_DMA_CHAN_EN,
-				      CHAN_MASK(channel), 0,
+	dw_write(dma, DW_CFG_LOW(channel), chan->cfg_lo | DW_CFG_CH_SUSPEND |
```
Comment (xiulipan) [inline src/drivers/dw/dma.c]: Please add the comment and explanation here. Maybe include the doc reference here too.
I think the inline comments here is more important for else to understand this. This looks like magic to me

File: src/drivers/dw/dma.c [hunk 4]
```diff
@@ -639,13 +639,19 @@ static int dw_dma_stop(struct dma *dma, int channel)
...
+
+	ret = poll_for_register_delay(dma_base(dma) + DW_CFG_LOW(channel),
+				      DW_CFG_CH_FIFO_EMPTY,
+				      DW_CFG_CH_FIFO_EMPTY,
 				      PLATFORM_DMA_TIMEOUT);
 	if (ret < 0)
 		trace_dwdma_error("dw-dma: %d channel %d timeout",
 				  dma->plat_data.id, channel);
 
+	dw_write(dma, DW_DMA_CHAN_EN, CHAN_DISABLE(channel));
```
Comment (xiulipan) [inline src/drivers/dw/dma.c]: As you mentioned above the dw-dma will auto stop after we wait. Why need to disable it again?

Comment (xiulipan) [inline src/drivers/dw/dma.c]: If the dma will auto stop, why we need to disable it here?

**Comment by juimonen** [comment]
@tlauda changed PLATFORM_DMA_TIMEOUT to 4000

Was this what you meant? It didn't help the timeout error in stop...

**Comment by tlauda** [comment]
@juimonen No, I meant IPC timeout on kernel side.

File: src/drivers/dw/dma.c (hunk 4 above)
Comment (tlauda) [reply src/drivers/dw/dma.c]: Everything is explained in commit message. Wihtout explicit stop you can potentially wait up to 1 ms, which for sure will cause IPC timeout.

File: src/drivers/dw/dma.c (hunk 3 above)
Comment (tlauda) [reply src/drivers/dw/dma.c]: Done.

**Comment by juimonen** [comment]
@tlauda ipc timeout in kernel side is 300ms... and dmesg also shows time between tx and timeout about 300 ms. Am I missing something here as I see it, kernel is waiting for DSP 300ms without answer...

**Comment by tlauda** [comment]
@juimonen If it's 300 ms, then we have additional problem in kernel with getting responses as @xiulipan stated in thesofproject/linux#808.

File: src/drivers/dw/dma.c (hunk 2 above)
Comment (xiulipan) [reply src/drivers/dw/dma.c]: @lyakh This means we need to keep the other bits to the original status as we only changes few bits

**Comment by xiulipan** [comment]
@tlauda @lyakh 
I can confirm that it is about 300 ms for the timeout. From the https://github.com/thesofproject/linux/issues/808 you can also find the timeout is about 300 ms.

**Pull request merged**